Processor and Component for iframes

iamuchejude opened the pull request:
#### Description of what you did:

Create a processor and component for iframes.

#### My PR is a:

- 🔝 Improvement

#### Is the PR ready to be merged?

- ✅ Yes!

#### Feature Request

https://community.frontity.org/t/lazy-load-iframes-processor/1116

#### Docs Changes

https://app.gitbook.com/@frontity-tech/s/frontity/~/drafts/-M-TwT82Z6C5MTL5s2-v/api-reference-1/frontity-html2react

----

Fixes #281.
Fixes #282.

iamuchejude committed:
Processor: Create processor for iframes

luisherranz commented on .vscode/launch.json:
```diff
@@ -0,0 +1,17 @@
+      "program": "${file}"
+    }
+  ]
+}
```
Hey Jude, don't commit your `.vscode` folder. You can add it to the `.gitignore` file if you want :)

iamuchejude replied:
Yea...I added it to my `.gitignore` in the script processor.

iamuchejude committed:
Component: write e2e test for iframe lazy-load component

iamuchejude committed:
Component: initial implementation of iframe component

luisherranz commented on lerna-debug.log:
```diff
@@ -0,0 +1,12 @@
+3 info versioning independent
+4 error JSONError: Unexpected end of JSON input while parsing near '' in packages/iframe/package.json
+4 error     at module.exports (/Users/jude/Projects/frontity/node_modules/parse-json/index.js:26:19)
+4 error     at parse (/Users/jude/Projects/frontity/node_modules/load-json-file/index.js:8:29)
```
This file has to be deleted. It's just an error log from lerna.

luisherranz commented on examples/mars-theme-example/frontity.settings.js:
```diff
@@ -32,7 +32,7 @@ export default {
       state: {
         source: {
-          api: "https://test.frontity.io/wp-json"
+          api: "http://frontity.local/wp-json"
```
Remember not to commit these changes.

iamuchejude replied:
Okay.

luisherranz commented on packages/mars-theme/src/components/index.js:
```diff
@@ -41,6 +42,31 @@ const Theme = ({ state }) => {
           (data.is404 && <Page404 />)}
       </Main>
+
+      <Iframe
```
Same here, remember not to commit these changes.

iamuchejude replied:
Sure

iamuchejude committed:
HTML2React: iframe processor

iamuchejude committed:
Component: Fix issues with the iframe component

iamuchejude committed:
HTML2React: Unit tests for iframe processor

luisherranz committed:
Merge branch 'dev' into iframe-processor

luisherranz committed:
Fix iframe component unit tests

luisherranz committed:
Fix iframe processor unit tests

luisherranz committed:
Add file for image processor unit tests

luisherranz committed:
Remove internal package-lock files

luisherranz commented:
This is the table that represents the current approach to lazy loading in case you want to add it to the documentation:

| Intersection Observer | Native Lazy | Height > 0 | Output                |
|-----------------------|-------------|--------|-----------------------|
| true                  | true        | true   | Native Lazy Load             |
| true                  | true        | false  | Intersection Observer |
| true                  | false       | true   | Intersection Observer |
| true                  | false       | false  | Intersection Observer |
| false                 | true        | true   | (not possible)                     |
| false                 | true        | false  | (not possible)                     |
| false                 | false       | true   | Normal Load (eager)            |
| false                 | false       | false  | Normal Load (eager)            |

luisherranz committed:
Merge branch 'dev' into iframe-processor

luisherranz committed:
Remove headless check in image e2e tests

Because now the headless version uses Chrome 78 and it includes
native lazy load.

luisherranz committed:
Remove headless check in iframe e2e tests

Because now the headless version uses Chrome 78 and it includes
native lazy load.

luisherranz commented:
@DAreRodz could you please help @iamuchejude with the e2e tests of the iframe component?

By the way, I've removed the "HEADLESS" check because the headless version now uses Chrome 78. So now both IO and native lazy load should pass on headless as well 👍

iamuchejude committed:
Update e2e tests for iframe component

iamuchejude committed:
Write unit tests for iframe component

iamuchejude committed:
HTML2React: Write unit tests for script

iamuchejude committed:
Update e2e test for Iframe component

DAreRodz commented:
After having a call with @iamuchejude, we saw that some of the tests for the component using `IntersectionObserver` were failing because of a tiny error in the implementation. It's fixed now and they are passing, so that's great.

Regarding the native lazy-loading, the tests are passing in the normal browser, let's see if they pass in the headless browser as well. @iamuchejude, can you check that running `npm run e2e:run` in `frontity/e2e`?

iamuchejude commented:
Okay, @DAreRodz. I'm on it.

luisherranz commented:
> Regarding the native lazy-loading, the tests are passing in the normal browser, let's see if they pass in the headless browser as well.

I updated Cypress and headless is now using Chrome 78 so they should pass.

Great work guys 👍

iamuchejude committed:
Components: Fix issue in the iframe component

iamuchejude committed:
Fix issues with e2e tests for the iframe component

iamuchejude committed:
Component: Update unit tests for the iframe component

iamuchejude committed:
Update e2e test for iframe component

iamuchejude committed:
Update unit tests for components

iamuchejude committed:
Merge branch 'dev' into iframe-processor

luisherranz committed:
Remove screenshot from failed e2e test

luisherranz committed:
Use ES5 syntax for old browsers script

luisherranz committed:
Fix bug in old browsers script class name

Without the space before the class name it becomes
"frontity-lazy-iframeother-class" instead of
"frontity-lazy-iframe other-class".

luisherranz committed:
Remove supressHydrationWarning

Not needed when we use state.frontity.rendering.

luisherranz committed:
Update component snapshots

luisherranz committed:
Change e2e test name

I was a bit confused at what to expect from that test.

iamuchejude committed:
Add changeset

luisherranz reviewed (approved):
Thanks @iamuchejude! I am going to merge this.

I've done some changes to your code thought, please take a look at my last commits:
- Added a changeset (remember to include one next time).
- Fixed a bug with classnames not having an space between them.
- Used ES5 syntax for the old-browser script.
- Remove screenshot from failed e2e test (don't commit those, although we should include this in the gitignore as well).
